### ISSUE TITLE
Which requests can be retried?

### ISSUE THREAD

**Issue by jakearchibald** [jakearchibald/background-fetch#26]
We need to think about which cases are "retryable", and which indicate terminal failure. Eg, a POST that results in a network failure may be "retryable", but a POST that results in a 403 may not. We also need to define how often a "retryable" request can be retried before failing, and any kind of delay before retrying.

Range requests probably don't make sense for non-GET requests.

**Comment by arturkulig** [comment]
In case of any valid response, bgFetchJob should be considered a success and trigger `backgroundfetched`. It is a matter of interpretation if non-200 response requires to be redone.
At times 404 could be a expected response, if a request has been made when other information that could help establish if the bgFetchJob is going to succeed, was not available.

Therefore `backgroundfetchfailed` should provide all informations that would help establish a reason of invalid HTTP response. `bgFetchEvent` could be extended, to provide information which transmission (lie-fi / dns / dropped) erroneous scenario occurred.

Dropped GETs and then usage of HTTP ranges would mean glueing multiple responses manually, which would be inconvenient, but not impossible.

**Comment by jakearchibald** [comment]
> In case of any valid response, bgFetchJob should be considered a success and trigger `backgroundfetched`. It is a matter of interpretation if non-200 response requires to be redone.
At times 404 could be a expected response, if a request has been made when other information that could help establish if the bgFetchJob is going to succeed, was not available.

We made this call with `cache.addAll` in the cache API, and it was not what developers were expecting. Although this is how `fetch()` behaves.

> Therefore `backgroundfetchfailed` should provide all informations that would help establish a reason of invalid HTTP response. `bgFetchEvent` could be extended, to provide information which transmission (lie-fi / dns / dropped) erroneous scenario occurred.

I'm hoping the fetch observer will give us this. I don't want to end up with two things doing the same job.

> Dropped GETs and then usage of HTTP ranges would mean glueing multiple responses manually, which would be inconvenient, but not impossible.

The HTTP cache already does this, and browser can already pause/resume downloads, so I think the logic's already there.

**Comment by arturkulig** [comment]
I'm very happy with how fetch treats failures signalled with HTTP codes.

> The HTTP cache already does this, and browser can already pause/resume downloads, so I think the logic's already there.

That is if you let browser handle that, but I was trying to justify not providing default nor automatic behaviour as it can be quite easily expressed with event handlers.

Then again I described it as inconvenient and, as you say, browsers know how to retry a download, so to reconcile both, retry strategy options could be helpful. However retrying using HTTP ranges sounds like something more adequate to single `Request` rather than whole group.

**Comment by jakearchibald** [comment]
> I was trying to justify not providing default nor automatic behaviour as it can be quite easily expressed with event handlers.

We can't really have pause/resume without it, unless we expect developers to roll their own here.

> However retrying using HTTP ranges sounds like something more adequate to single `Request` rather than whole group.

Pausing a background fetch may pause multiple in-flight fetches.

**Comment by arturkulig** [comment]
I assumed that pausing is about discontinuing making new requests (in case of many fetches).
Also - chunking upload requires more than HTTP, doesn't it?

I'd expect:
- pausing upload to drop connection (**as default** behaviour)
- pausing download to discard downloaded content (**as default** behaviour)
- HTTP byte range usage as optional

With pausing and resuming later also comes a danger of remote resource changing meantime (as [immutable responses](http://httpwg.org/http-extensions/immutable.html) are not a thing yet), which would made byte range in that case useless.

**Comment by jakearchibald** [comment]
> I assumed that pausing is about discontinuing making new requests

No, it needs to stop using data as soon as possible. The user may be pausing because they've started roaming, or intend to start. If I pause something but progress continues - that seems broken.

> Also - chunking upload requires more than HTTP, doesn't it?

Right. I've mentioned this in the explainer.

> I'd expect pausing download to discard downloaded content

This sounds like stopping, not pausing.

> With pausing and resuming later also comes a danger of remote resource changing meantime

I think the HTTP spec covers this somewhat. We shouldn't combine partial responses if the etag or last-modified changes. I guess we can't resume if neither of these headings is present.

**Comment by arturkulig** [comment]
upload => 👍  sure thing, you have

> Eventually we'd like uploads to be resumable, but there's no standard for that.

What I was initially about, was to suggest giving enough information in error messages, so dev can implement such protocol of chunked/resumed upload their service supports. Currently `backgroundfetchfailed` event object has same shape as `backgroundfetched`, where I don't see how either `Response` or `Request` object would pass information why request failed.
(`Fetch` is also not returning very useful errors when raising requests while i.e. offline.)

download => 👍  You're absolutely right - this behaviour can be conditional in way you described it and still very much reliable.
